client/failing_funding_test.go

```diff
@@ -0,0 +1,102 @@
+// Copyright 2022 - See NOTICE file for copyright holders.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package client_test
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"math/big"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	"perun.network/go-perun/channel"
+	chtest "perun.network/go-perun/channel/test"
+	"perun.network/go-perun/client"
+	ctest "perun.network/go-perun/client/test"
+	"perun.network/go-perun/wire"
+	"polycry.pt/poly-go/test"
+)
+
+const (
+	fridaIdx     = 0
+	fredIdx      = 1
+	fridaInitBal = 100
+	fredInitBal  = 50
+)
+
+func TestFailingFunding(t *testing.T) {
+	ctx, cancel := context.WithTimeout(context.Background(), twoPartyTestTimeout)
+	defer cancel()
+	rng := test.Prng(t)
+
+	setups := NewSetups(rng, []string{"Frida", "Fred"})
+	// Inject the failing funder for Frida.
+	setups[fridaIdx].Funder = FailingFunder{}
+	clients := NewClients(t, rng, setups)
+	frida, fred := clients[fridaIdx], clients[fredIdx]
+	fridaAddr, fredAddr := frida.Identity.Address(), fred.Identity.Address()
+
+	chsFred := make(chan *client.Channel, 1)
+	errsFred := make(chan error)
+	go fred.Handle(
+		ctest.AlwaysAcceptChannelHandler(ctx, fredAddr, chsFred, errsFred),
+		ctest.AlwaysRejectUpdateHandler(ctx, errsFred),
+	)
+
+	// Create the proposal.
+	asset := chtest.NewRandomAsset(rng)
+	initAlloc := channel.NewAllocation(2, asset)
+	initAlloc.SetAssetBalances(asset, []*big.Int{big.NewInt(fridaInitBal), big.NewInt(fredInitBal)})
+	parts := []wire.Address{fridaAddr, fredAddr}
+	prop, err := client.NewLedgerChannelProposal(
+		challengeDuration,
+		fridaAddr,
+		initAlloc,
+		parts,
+	)
+	require.NoError(t, err, "creating ledger channel proposal")
+
+	// Frida sends the proposal.
+	_, err = frida.ProposeChannel(ctx, prop)
+	// We expect a ChannelFunding
+	cfErr, ok := err.(*client.ChannelFundingError)
+	require.Truef(t, ok, fmt.Sprintf("expexted ChannelFundingError, got %T", err))
+	require.Nil(t, cfErr.SettleError)
+
+	// Fred gets the channel and settle it afterward.
+	var ch *client.Channel
+	select {
+	case ch = <-chsFred:
```

Comment:
You mean that when the funding fails for one party, it should consequently fail for the other party, right? As this should be the case with a real backend.
I'll look into how I could simulate this.

Reply:
It must fail for all parties. Otherwise the tests would be tailored to a not correctly functioning backend, but they should work for all correctly implemented backends. One way to do this is to implement the asset holder functionality on the mock backend. That is, track which funds have been provided for a channel. Then `backend.Fund` fails if not all funds have been provided within a given timeout.

Reply:
I now implemented this behavior. :D